libpod/image/image.go

```diff
@@ -244,48 +240,143 @@ func (ir *Runtime) SaveImages(ctx context.Context, namesOrIDs []string, format s
 			if isTagged {
 				iData.tags = append(iData.tags, refTagged)
 			}
+		} else {
+			iData.destNames = []string{""}
 		}
 	}
 
-	policyContext, err := getPolicyContext(sys)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := policyContext.Destroy(); err != nil {
-			logrus.Errorf("failed to destroy policy context: %q", err)
+	switch format {
+	case DockerArchive:
+		archWriter, err := archive.NewWriter(sys, outputFile)
+		if err != nil {
+			return err
 		}
-	}()
+		defer func() {
+			err := archWriter.Close()
+			if err == nil {
+				return
+			}
+			if finalErr == nil {
+				finalErr = err
+				return
+			}
+			finalErr = errors.Wrap(finalErr, err.Error())
+		}()
+		// Now copy the images one-by-one.
+		for _, id := range imageQueue {
+			dest, err := archWriter.NewReference(nil)
+			if err != nil {
+				return err
+			}
+
+			img := imageMap[id]
+			copyOptions := getCopyOptions(sys, writer, nil, nil, SigningOptions{RemoveSignatures: removeSignatures}, "", img.tags)
+			copyOptions.DestinationCtx.SystemRegistriesConfPath = registries.SystemRegistriesConfPath()
 
-	// Now copy the images one-by-one.
-	for _, id := range imageQueue {
-		dest, err := archWriter.NewReference(nil)
+			// For copying, we need a source reference that we can create
+			// from the image.
+			src, err := is.Transport.NewStoreReference(img.imageruntime.store, nil, id)
+			if err != nil {
+				return errors.Wrapf(err, "error getting source imageReference for %q", img.InputName)
+			}
+			_, err = cp.Image(ctx, policyContext, dest, src, copyOptions)
+			if err != nil {
+				return err
+			}
+		}
+	case OCIArchive:
+		archWriter, err := ociarchive.NewWriter(sys, outputFile)
 		if err != nil {
 			return err
 		}
+		defer func() {
+			err := archWriter.Close()
+			if err == nil {
+				return
+			}
+			if finalErr == nil {
+				finalErr = err
+			}
+			finalErr = errors.Wrap(finalErr, err.Error())
+		}()
 
-		img := imageMap[id]
-		copyOptions := getCopyOptions(sys, writer, nil, nil, SigningOptions{RemoveSignatures: removeSignatures}, "", img.tags)
+		copyOptions := getCopyOptions(sys, writer, nil, nil, SigningOptions{RemoveSignatures: removeSignatures}, "", nil)
 		copyOptions.DestinationCtx.SystemRegistriesConfPath = registries.SystemRegistriesConfPath()
+		// copy the image one by one
+		for _, id := range imageQueue {
+			img := imageMap[id]
+			// For copying, we need a source reference that we can create
+			// from the image.
+			src, err := is.Transport.NewStoreReference(img.imageruntime.store, nil, id)
```

Comment:
This lookup can AFAICS also happen in the shared `imageQueue` construction code.

Reply:
The src,  `storageReference` is not exported. The imageQueue may not be able to hold a field for it.